python/lsst/daf/butler/core/dimensions/storage.py

```diff
@@ -0,0 +1,522 @@
+        if element.hasTable() and element.spatial and commonSkyPixOverlapTable is None:
+            raise TypeError(f"No common skypix table provided for element {element.name}.")
+
+    @property
+    def element(self) -> DimensionElement:
+        # Docstring inherited from DimensionRecordStorage.element.
+        return self._element
+
+    def clearCaches(self):
+        # Docstring inherited from DimensionRecordStorage.clearCaches.
+        pass
+
+    def matches(self, dataId: Optional[DataId] = None) -> bool:
+        # Docstring inherited from DimensionRecordStorage.matches.
+        return True
+
+    def getElementTable(self, dataId: Optional[DataId] = None) -> FromClause:
+        # Docstring inherited from DimensionRecordStorage.getElementTable.
+        return self._elementTable
+
+    def getCommonSkyPixOverlapTable(self, dataId: Optional[DataId] = None) -> FromClause:
+        # Docstring inherited from
+        # DimensionRecordStorage.getCommonSkyPixOverlapTable.
+        return self._commonSkyPixOverlapTable
+
+    def insert(self, *records: DimensionRecord):
+        # Docstring inherited from DimensionRecordStorage.insert.
+        if self.element.viewOf is not None:
+            raise TypeError(f"Cannot insert {self.element.name} records because its "
+                            f"logical table is a view into {self.element.viewOf}.")
+        # Build lists of dicts to insert first, before any database operations,
+        # to minimize the time spent in the transaction.
+        elementRows = []
+        if self.element.spatial:
+            commonSkyPixRows = []
+            commonSkyPix = self.element.universe.commonSkyPix
+        for record in records:
+            elementRows.append(record.toDict())
+            if self.element.spatial:
+                if record.region is None:
+                    # TODO: should we warn about this case?
+                    continue
+                base = record.dataId.byName()
+                for begin, end in commonSkyPix.pixelization.envelope(record.region):
+                    for skypix in range(begin, end):
+                        row = base.copy()
+                        row[commonSkyPix.name] = skypix
+                        commonSkyPixRows.append(row)
+        # TODO: wrap the operations below in a transaction.
+        self._connection.execute(self._elementTable.insert(), *elementRows)
+        if self.element.spatial and commonSkyPixRows:
+            self._connection.execute(self._commonSkyPixOverlapTable.insert(), *commonSkyPixRows)
+
+    def fetch(self, dataId: DataCoordinate) -> Optional[DimensionRecord]:
+        # Docstring inherited from DimensionRecordStorage.fetch.
+        RecordClass = self.element.RecordClass
+        # I don't know how expensive it is to construct the query below, and
+        # hence how much gain there might be to caching it, so I'm going to
+        # wait for it to appear as a hotspot in a profile before trying that.
```

Comment:
There's a trade-off here between multi-row select, caching, and fetching duplicate values from the database, so at some level structuring fetch as a single-record operation is a guess that caching and avoiding fetching duplicates is more valuable for performance than multi-row select.  I think there's some indirect evidence for that being true, at least the preflight queries, from the last round of optimizing we did there.

Reply:
Hmm, I would expect multi-row select (with `multi >> 1`) to produce bigger benefit simply because of fewer round-trips to database. But indeed we need it to be measured first to understand if it's an issue.